### DISCUSSION TITLE
using knex migrations and seed

### DISCUSSION THREAD

**Discussion by codelover2k** [oguimbal/pg-mem#106] [Q&A]
hi!
I am using knex with multiple migration files.

I know that you can use knex with pg-mem, but is there a way to tell pg-meme to use knex and create the database with the multiple migration files and use knex seed functionality all in one step for each test?

A simple example would be great. Thanks!

**Comment by oguimbal** [comment]
Hi,

There is no difference between an instance of Knex you would create as usual, and an instance of Knex [you can create through pg-mem](https://github.com/oguimbal/pg-mem/blob/f3c01089aa540e8c7938424456f2478523a6f110/samples/knex/knex.ts#L13).

I dont know Knex enough to answer the question "how to do it ?" (I dont know how to use migration files with it), but I can answer: "Like you would with Knex"... I suppose you do that through Knex configuration options, so:

```typescript
const pgMem = newDb();
const knex = db.adapters.createKnex(0, {  /* PUT YOUR KNEX OPTIONS HERE */ });
```

**Reply by codelover2k** [reply]
thanks for response!
I try to figure something out.

Beside of that, my API uses postgres, is there a way to tell my server to use pg-mem in testing mode?

With other words I am not using the database directly, I am using the service and after I am done with the test, I can create a new database instance that my server can then use.

**Reply by oguimbal** [reply]
I'm not sure to understand your setup... what do you mean by "not using the database directly" ?

**Reply by codelover2k** [reply]
if you use pg-mem, its like using your database/ORM/query builder directly.

Saying that, lets say you have something like this:

[APP/API] -------> [DATABASEDRIVER] -------> [DATABASE]

now I want to use pg-mem and replace it with [DATABASE], with other words, I would like that my app, uses pg-mem as the database to be able to test my API

**Reply by oguimbal** [reply]
Currently, the only way you have to do that is to hook your database driver to override how the queries are performed when you create the connection to your database (= when you configure your ORM instance), in order to avoid your ORM performing actual network calls.

That is the purpose of the different adapters.

If you'd like pg-mem to act as a database instance that is reachable like a real database (i.e. over the network via the pg protocol), then you will have to wait until I integrate [pg-server](https://github.com/oguimbal/pg-server) with pg-mem, which you can track via [this issue](https://github.com/oguimbal/pg-mem/issues/27).

... that is a less performant solution, though, since it involves networking.

**Reply by codelover2k** [reply]
will wait for it, meanwhile I am going to use docker with tempfs.
thanks!

**Comment by capan** [comment]
I have a similar question;
I was able to run my migrations using pg-mem but in my case I'm receiving SQL syntax error during the execution of my migration files.

The error says;

```
'alter table "business" add column "iban" text, add column "deposit_period" integer - 💔 Your query failed to parse.
This is most likely due to a SQL syntax error. However, you might also have hit a bug, or an unimplemented feature of pg-mem.
If this is the case, please file an issue at https://github.com/oguimbal/pg-mem along with a query that reproduces this syntax error.

👉 Failed query:

    alter table "business" add column "iban" text, add column "deposit_period" integer;

💀 Syntax error at line 1 col 46:

  alter table "business" add column "iban" text,
                                               ^
Unexpected comma token: ",". Instead, I was expecting to see one of the following:

    - A "dot" token
    - A "lparen" token
    - A "kw_array" token
    - A "lbracket" token
    - A "kw_collate" token
    - A "kw_constraint" token
    - A "kw_unique" token
    - A "kw_null" token
    - A "kw_default" token
    - A "kw_check" token
    - A "kw_primary" token
    - A "kw_not" token
    - A "word" token
    - A "semicolon" token'
```

What is wrong with this SQL query?

**alter table "business" add column "iban" text, add column "deposit_period" integer**

**Reply by oguimbal** [reply]
This is on my todo list... the parser does not support parsing ALTER statements that are altering more than one columns. See [this issue](https://github.com/oguimbal/pgsql-ast-parser/issues/57)
